### PR TITLE
Gen info and selection

### PR DESCRIPTION
**Pull request by abolshov**
Implementation of searching for signal events and gen level objects for both SL and DL;
Tested on two samples (1 DL sample and 1 SL sample), no crashes observed.

### CODE REVIEW

File: include/BaselineGenSelection.h [hunk 1]
```diff
@@ -5,11 +5,297 @@
...
+                                      [&](int idx){ return std::abs(GenPart_pdgId[idx]) <= PdG::b(); });
+    bool v1_to_leptons = std::all_of(daughters_v1.begin(), daughters_v1.end(),
+                                      [&](int idx){ return PdG::isNeutrino(GenPart_pdgId[idx]) || PdG::isLepton(GenPart_pdgId[idx]); });
+
+    RVecI daughters_v2 = GenPart_daughters.at(v2_idx);
+    bool v2_to_hadrons = std::all_of(daughters_v2.begin(), daughters_v2.end(),
+                                      [&](int idx){ return std::abs(GenPart_pdgId[idx]) <= PdG::b(); });
+    bool v2_to_leptons = std::all_of(daughters_v2.begin(), daughters_v2.end(),
+                                      [&](int idx){ return PdG::isNeutrino(GenPart_pdgId[idx]) || PdG::isLepton(GenPart_pdgId[idx]); });
+
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:196]: do it in a loop

File: include/BaselineGenSelection.h [hunk 2]
```diff
@@ -5,11 +5,297 @@
...
+}
+
+HBBCand GetGenHBBCandidate(int evt, const RVecI& GenPart_pdgId, const RVecVecI& GenPart_daughters, const RVecI& GenPart_statusFlags,
+                           const RVecF& GenPart_pt, const RVecF& GenPart_eta, const RVecF& GenPart_phi, const RVecF& GenPart_mass,
+                           bool throw_error_if_not_found)
+{
+  try
+  {
+    int sz = GenPart_pdgId.size();
+    std::set<int> Hbb_indices;
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:238]: can `GetGenHBBIndex` be reused?

**Review by kandrosov — CHANGES_REQUESTED** [review]
also, please, fix merge conflicts.

File: include/BaselineGenSelection.h [hunk 3]
```diff
@@ -5,11 +5,297 @@
...
+    cand.leg_index[1] = daughters[1];
+    cand.leg_kind[1] = static_cast<Vleg>(l2_ptr->kind());
+    cand.leg_p4[1] = l2_ptr->lastCopy().p4;
+    cand.leg_vis_p4[1] = LorentzVectorM(l2_ptr->visibleP4());
+
+    return;
+  }
+
+  int nu_idx = daughters.at(1);
+  int lep_idx = daughters.at(0);
```
[the remark this answers is not in the document]
Comment (abolshov) [reply include/BaselineGenSelection.h:85]: In the beginning of this function vector of daughters is sorted by absolute value of pdg id. Absolute value of pdg id of neutrino is always greater than pdg id of lepton. So lepton will be first (i.e. at index 0) and neutrino will be second (i.e. at index 1).

File: include/BaselineGenSelection.h (hunk 1 above)
Comment (abolshov) [reply include/BaselineGenSelection.h:196]: Access to daughters of both vector bosons simultaneously is necessary check that the H->VV decays belongs to double lepton channel or single lepton channel. It is impossible to implement it in a loop.

File: include/BaselineGenSelection.h [hunk 4]
```diff
@@ -5,11 +5,303 @@
 #include "TextIO.h"
 #include "HHCore.h"
 
+
+void AssignHadronicVCand(VCand& cand, int cand_idx, const RVecI& GenPart_pdgId, const RVecVecI& GenPart_daughters,
+                         const RVecF& GenPart_pt, const RVecF& GenPart_eta, const RVecF& GenPart_phi, const RVecF& GenPart_mass,
+                         RVecLV const& GenJet_p4)
+{
+  cand.leg_kind[0] = Vleg::Jet;
+  cand.leg_kind[1] = Vleg::Jet;
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:14]:
```suggestion
```

File: include/BaselineGenSelection.h [hunk 5]
```diff
@@ -5,11 +5,303 @@
...
+  cand.cand_p4 = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, cand_idx);
+
+  cand.leg_p4[0] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, ParticleDB::GetMass(GenPart_pdgId[cand.leg_index[0]], GenPart_mass[cand.leg_index[0]]), cand.leg_index[0]);
+  cand.leg_p4[1] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, ParticleDB::GetMass(GenPart_pdgId[cand.leg_index[1]], GenPart_mass[cand.leg_index[1]]), cand.leg_index[1]);
+
+  int match_1 = FindMatching(cand.leg_p4[0], GenJet_p4, 0.4);
+  int match_2 = FindMatching(cand.leg_p4[1], GenJet_p4, 0.4);
+
+  cand.leg_vis_p4[0] = match_1 == -1 ? LorentzVectorM{} : GenJet_p4[match_1];
+  cand.leg_vis_p4[1] = match_2 == -1 ? LorentzVectorM{} : GenJet_p4[match_2];
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:31]:
```suggestion
  for(size_t leg_id = 0; leg_id < cand.leg_p4.size(); ++leg_id) {
    cand.leg_kind[leg_id] = Vleg::Jet;
    const int leg_idx = cand.leg_index[leg_id];
    const double mass = ParticleDB::GetMass(GenPart_pdgId[leg_idx], GenPart_mass[leg_idx]);
    cand.leg_p4[leg_id] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, mass, leg_idx);
    const int match = FindMatching(cand.leg_p4[leg_id], GenJet_p4, 0.4);
    cand.leg_vis_p4[leg_id] = match_1 == -1 ? LorentzVectorM{} : GenJet_p4[match];
  }
```

File: include/BaselineGenSelection.h [hunk 6]
```diff
@@ -5,11 +5,303 @@
...
+
+  cand.leg_index[1] = nu_idx;
+  cand.leg_kind[1] = Vleg::Nu;
+  cand.leg_p4[1] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, nu_idx);
+  cand.leg_vis_p4[1] = LorentzVectorM(0, 0, 0, 0);
+
+  cand.leg_index[0] = lep_idx;
+  cand.leg_kind[0] = static_cast<Vleg>(lep_ptr->kind());
+  cand.leg_p4[0] = lep_ptr->lastCopy().p4;
+  cand.leg_vis_p4[0] = LorentzVectorM(lep_ptr->visibleP4());
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:105]:
```suggestion
  for(size_t lep_id = 0; lep_id < cand.leg_p4.size(); ++lep_id) {
    cand.leg_index[lep_id] = daughters[lep_id];
    if(PdG::isNeutrino(GenPart_pdgId[daughters[lep_id]])) {
      cand.leg_kind[lep_id] = Vleg::Nu;
      cand.leg_p4[lep_id] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, daughters[lep_id]);
      cand.leg_vis_p4[lep_id] = LorentzVectorM(0, 0, 0, 0);
    } else {
      reco_tau::gen_truth::GenLepton const* lep_ptr = findLeptonByIndex(gen_leptons, daughters[lep_id]);
      if (!lep_ptr)
        throw analysis::exception("Could not find gen lepton by index");
      cand.leg_kind[lep_id] = static_cast<Vleg>(lep_ptr->kind());
      cand.leg_p4[lep_id] = lep_ptr->lastCopy().p4;
      cand.leg_vis_p4[lep_id] = LorentzVectorM(lep_ptr->visibleP4());
    }
  }
```

File: include/BaselineGenSelection.h [hunk 7]
```diff
@@ -5,11 +5,303 @@
...
+                                      GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, GenJet_p4);
+
+    if (single_lepton)
+    {
+      // let leg 0 be always W->qq and leg 1 leg be W->lv
+      if (HVV_cand.legs.at(0).leg_kind.at(0) != Vleg::Jet)
+      {
+        std::swap(HVV_cand.legs.at(0), HVV_cand.legs.at(1));
+      }
+    }
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:223]:
```suggestion
    const RVecI& V_from_H = GenPart_daughters.at(HVV_index);
    size_t n_leptons = 0;
    for(size_t v_id = 0; v_id < HVV_cand.legs.size(); ++v_id)
    {
      const int v_idx = GetLastCopy(V_from_H.at(v_id), GenPart_pdgId, GenPart_statusFlags, GenPart_daughters);
      HVV_cand.legs.at(v_id) = GetGenVCand(evt, v_idx, gen_leptons, GenPart_pdgId, GenPart_daughters,
                                           GenPart_statusFlags, GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass,
                                           GenJet_p4);
      for(size_t leg_id = 0; leg_id < HVV_cand.legs.at(v_id).leg_p4.size(); ++leg_id) {
        if(HVV_cand.legs.at(v_id).leg_kind[leg_id] != Vleg::Jet) {
          ++n_leptons;
        }
      }
    }
    if(n_leptons != 2 && n_leptons != 4)
      throw analysis::exception("Encountered decay neither in single lepton nor double lepton channel");

    if(n_leptons == 2 && HVV_cand.legs.at(0).leg_kind.at(0) != Vleg::Jet)
      std::swap(HVV_cand.legs.at(0), HVV_cand.legs.at(1));
```

File: include/BaselineGenSelection.h [hunk 8]
```diff
@@ -5,11 +5,303 @@
...
+    Hbb_cand.leg_index[0] = b_from_H.at(first);
+    Hbb_cand.leg_index[1] = b_from_H.at(second);
+
+    Hbb_cand.leg_p4[0] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, Hbb_cand.leg_index[0]);
+    Hbb_cand.leg_p4[1] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, Hbb_cand.leg_index[1]);
+
+    Hbb_cand.leg_vis_p4[0] = LorentzVectorM{};
+    Hbb_cand.leg_vis_p4[1] = LorentzVectorM{};
+
+    return Hbb_cand;
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:288]:
```suggestion
    int const Hbb_index = GetGenHBBIndex(evt, GenPart_pdgId, GenPart_daughters, GenPart_statusFlags);
    HBBCand Hbb_cand;
    Hbb_cand.cand_p4 = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, Hbb_index);

    const RVecI& b_from_H = GenPart_daughters.at(Hbb_index);
    const int first = GenPart_pt[b_from_H[0]] > GenPart_pt[b_from_H[1]] ? 0 : 1;
    const int second = (first + 1) % 2;
    Hbb_cand.leg_index[0] = b_from_H.at(first);
    Hbb_cand.leg_index[1] = b_from_H.at(second);

    for(size_t leg_id = 0; leg_id < Hbb_cand.leg_p4.size(); ++leg_id) {
      const int leg_idx = Hbb_cand.leg_index[leg_id];
      Hbb_cand.leg_p4[leg_id] = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, leg_idx);
      const int match = FindMatching(Hbb_cand.leg_p4[leg_id], GenJet_p4, 0.4);
      Hbb_cand.leg_vis_p4[leg_id] = match == -1 ? LorentzVectorM{} : GenJet_p4[match];
    }

    return Hbb_cand;
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/HH_bbWW/anaTupleDef.py [hunk 9]
```diff
@@ -80,156 +80,79 @@ def getDefaultColumnsToSave(isData):
...
+        for b_quark in [1, 2]:
+            for var in PtEtaPhiM:
+                name = f"genb{b_quark}_{var}"
+                dfw.DefineAndAppend(name, f"H_to_bb.leg_p4[{b_quark - 1}].{var}()")
+    elif not isData:
+        # save gen leptons matched to reco leptons
+        for lep in [1, 2]:
+            name = f"matchedGenLep{lep}"
+            dfw.Define(f"{name}_idx", f"MatchGenLepton(lep{lep}_p4, genLeptons, 0.4)")
+            dfw.Define(f"{name}_p4", f"return {name}_idx == -1 ? LorentzVectorM() : GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, idx)")
```
Comment (kandrosov) [inline AnaProd/HH_bbWW/anaTupleDef.py:151]: use visible gen lepton p4 (to correctly get tau decay cases)

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/anaTupleProducer.py [hunk 10]
```diff
@@ -146,15 +146,15 @@ def createAnatuple(inFile, treeName, outDir, setup, sample_name, anaCache, snaps
...
-        #     report.Print()
+    # for index,fileName in enumerate(outfilesNames):
+    #     outputRootFile= ROOT.TFile(fileName, "UPDATE", "", compression_settings)
+    #     rep = ReportTools.SaveReport(reports[index].GetValue(), reoprtName=f"Report")
+    #     outputRootFile.WriteTObject(rep, f"Report", "Overwrite")
+    #     if index==0:
+    #         outputRootFile.WriteTObject(hist_time, f"runtime", "Overwrite")
+    #     outputRootFile.Close()
+    #     # if print_cutflow:
+    #     #     report.Print()
```
[the remark this answers is not in the document]
Comment (kandrosov) [reply AnaProd/anaTupleProducer.py:157]: Do you have any issues with running the code if you undo this modification?

File: AnaProd/HH_bbWW/anaTupleDef.py [hunk 11]
```diff
@@ -80,156 +80,80 @@ def getDefaultColumnsToSave(isData):
...
+            dfw.DefineAndAppend(f"{name}_pdgId", f"GenPart_pdgId[ H_to_VV.legs[{boson - 1}].index ]")
+            for var in PtEtaPhiM:
+                dfw.DefineAndAppend(f"{name}_{var}", f"H_to_VV.legs[{boson - 1}].cand_p4.{var}()")
+
+        # save gen level products of vector boson decays (prod - index of product (quark, leptons or neutrinos))
+        for boson in [1, 2]:
+            for prod in [1, 2]:
+                name = f"genV{boson}prod{prod}"
+                for var in PtEtaPhiM:
+                    dfw.DefineAndAppend(f"{name}_{var}", f"H_to_VV.legs[{boson - 1}].leg_p4[{prod - 1}].{var}()")
```
Comment (kandrosov) [inline AnaProd/HH_bbWW/anaTupleDef.py:132]: also store visible p4

File: AnaProd/HH_bbWW/anaTupleDef.py [hunk 12]
```diff
@@ -80,156 +80,80 @@ def getDefaultColumnsToSave(isData):
...
+        # save gen level H->bb
+        dfw.Define("H_to_bb", """GetGenHBBCandidate(event, GenPart_pdgId, GenPart_daughters, GenPart_statusFlags, GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, GenJet_p4, true)""")
+        for var in PtEtaPhiM:
+            dfw.DefineAndAppend(f"genHbb_{var}", f"H_to_bb.cand_p4.{var}()")
+
+        # save gen level b quarks
+        for b_quark in [1, 2]:
+            for var in PtEtaPhiM:
+                name = f"genb{b_quark}_{var}"
+                dfw.DefineAndAppend(name, f"H_to_bb.leg_p4[{b_quark - 1}].{var}()")
```
Comment (kandrosov) [inline AnaProd/HH_bbWW/anaTupleDef.py:145]: also store visible p4

File: AnaProd/HH_bbWW/anaTupleDef.py [hunk 13]
```diff
@@ -80,156 +80,80 @@ def getDefaultColumnsToSave(isData):
...
+    elif not isData:
+        # save gen leptons matched to reco leptons
+        for lep in [1, 2]:
+            name = f"matchedGenLep{lep}"
```
Comment (kandrosov) [inline AnaProd/HH_bbWW/anaTupleDef.py:149]:
```suggestion
            name = f"lep{lep}_genLep"
```

File: config/HH_bbWW/Run3_2022/samples.yaml [hunk 14]
```diff
@@ -1,6 +1,18 @@
 
-GluGlutoRadiontoHHto2B2Vto2B2JLNu_M-450:
-  sampleType: GluGluToRadion
-  mass: 260
+# GluGlutoRadiontoHHto2B2Vto2B2JLNu_M-450:
```
Comment (kandrosov) [inline config/HH_bbWW/Run3_2022/samples.yaml:2]: undo modifications in this file (it has been updated in the other PR)

File: config/Run3_2022/samples.yaml [hunk 15]
```diff
@@ -7,11 +7,11 @@ GLOBAL:
   crossSectionsFile : config/crossSections13TeV.yaml
   use_stitching: false
 
-DYto2L_M-50_madgraphMLM:
```
Comment (kandrosov) [inline config/Run3_2022/samples.yaml:10]: undo modifications in this file

File: include/BaselineGenSelection.h [hunk 16]
```diff
@@ -5,11 +5,297 @@
...
+      HVV_cand.legs.at(v_id).index = v_idx;
+      for(size_t leg_id = 0; leg_id < HVV_cand.legs.at(v_id).leg_p4.size(); ++leg_id) {
+        if(HVV_cand.legs.at(v_id).leg_kind[leg_id] != Vleg::Jet) {
+          ++n_leptons;
+        }
+      }
+    }
+    if(n_leptons != 2 && n_leptons != 4)
+      throw analysis::exception("Encountered decay neither in single lepton nor double lepton channel");
+    if(n_leptons == 2 && HVV_cand.legs.at(0).leg_kind.at(0) != Vleg::Jet)
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:144]: I think we want the first V to be leptonic, or not?

File: include/BaselineGenSelection.h [hunk 17]
```diff
@@ -5,11 +5,297 @@
...
+  {
+    if(throw_error_if_not_found)
+    {
+      throw analysis::exception("GetGenHVVCandidate (event=%1%): %2%") % evt % e.message();
+    }
+    return {};
+  }
+}
+
+// OLD BEGIN
```
Comment (kandrosov) [inline include/BaselineGenSelection.h:159]: remove old code

**Review by kandrosov — CHANGES_REQUESTED** [review]
Please apply the remaining comments and resolve merge conflicts.

File: AnaProd/HH_bbWW/anaTupleDef.py [hunk 18]
```diff
@@ -80,156 +80,83 @@ def getDefaultColumnsToSave(isData):
...
+                dfw.DefineAndAppend(f"{name}_pdgId", f"GenPart_pdgId[ H_to_VV.legs.at({boson - 1}).leg_index.at({prod - 1}) ]")
+
+        # save gen level H->bb
+        dfw.Define("H_to_bb", """GetGenHBBCandidate(event, GenPart_pdgId, GenPart_daughters, GenPart_statusFlags, GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, GenJet_p4, true)""")
+        for var in PtEtaPhiM:
+            dfw.DefineAndAppend(f"genHbb_{var}", f"H_to_bb.cand_p4.{var}()")
+
+        # save gen level b quarks
+        for b_quark in [1, 2]:
+            name = f"genb{b_quark}"
```
Comment (kandrosov) [inline AnaProd/HH_bbWW/anaTupleDef.py:144]:
```suggestion
```

**Review by kandrosov — APPROVED** [review]

**Pull request merged**